### PR TITLE
Update to latest cron lib.

### PR DESCRIPTION
**Pull request by artursouza**
# Description

Update to latest cron lib.

## Issue reference

Please reference the issue this PR will close: N/A
## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by artursouza** [comment]
I verified that ITs are passing with: `go test -v -race -tags integration ./tests/integration -focus scheduler`

File: dapr/proto/runtime/v1/dapr.proto
Comment (artursouza) [inline dapr/proto/runtime/v1/dapr.proto:199]: We won't support these APIs until there is an use case.

File: dapr/proto/scheduler/v1/scheduler_callback.proto
Comment (artursouza) [inline dapr/proto/scheduler/v1/scheduler_callback.proto:1]: This is not needed anymore with the streaming solution being implemented by @cicoyle.

File: pkg/runtime/config.go
Comment (artursouza) [inline pkg/runtime/config.go:98]: These changes are for consistency with the other flags. If we want to change this, we should have a separate PR making it all consistent.

File: tests/integration/framework/util/parallel.go
Comment (artursouza) [inline tests/integration/framework/util/parallel.go]: @JoshVanL This method is now deprecated and it just panics. Is this needed still?

File: cmd/daprd/options/options.go
Comment (cicoyle) [inline cmd/daprd/options/options.go]:
```suggestion
	fs.StringVar(&opts.SchedulerAddress, "scheduler-host-address", "", "Addresses of the Scheduler service instance(s), as comma separated host:port pairs")
```

File: pkg/actors/actors.go
Comment (cicoyle) [inline pkg/actors/actors.go]: the request period is the schedule?

File: pkg/scheduler/server/server.go
Comment (cicoyle) [inline pkg/scheduler/server/server.go:156]: I liked this as is, as `runEtcd` since its starting the embedded etcd first, then initializes the etcd cron instance,  configures the etcd-cron with an etcd client, and starts the etcd cron scheduler. 

I think `runEtcd` encapsulates both parts

File: tests/integration/suite/scheduler/api/http/errors.go
Comment (cicoyle) [inline tests/integration/suite/scheduler/api/http/errors.go:259]: I think we still want this

**Review by cicoyle — APPROVED** [review]
Minor comments (plz address), otherwise looks good 👍🏻

File: pkg/actors/actors.go
Comment (artursouza) [reply pkg/actors/actors.go]: I was thinking of sending it as-is and letting the library handle period too. Now, thinking about it, I will put the "@Every" back. So, I can extend the format for "@Every x"

File: pkg/scheduler/server/server.go
Comment (artursouza) [reply pkg/scheduler/server/server.go:156]: We will support running EtcdCron without Etcd to allow a multi-tenant Etcd instance. When I was reading this, I though it was just Etcd.

Comment (artursouza) [reply pkg/scheduler/server/server.go:156]: @JoshVanL any thoughts on this?

File: tests/integration/suite/scheduler/api/http/errors.go
Comment (artursouza) [reply tests/integration/suite/scheduler/api/http/errors.go:259]: The current implementation, deleting an object that is not found is also a success. So, there is no error to test.

File: dapr/proto/runtime/v1/dapr.proto
Comment (JoshVanL) [reply dapr/proto/runtime/v1/dapr.proto:199]: Please do not remove these APIs. They are invaluable to applications for discovery and "ensure" type operations. They are also part of the original proposal https://github.com/dapr/proposals/pull/44
I don't see any need to be removing these.

File: pkg/actors/actors.go
Comment (JoshVanL) [inline pkg/actors/actors.go:1217]: Why is the app ID being removed from the name?

File: pkg/runtime/config.go
Comment (JoshVanL) [reply pkg/runtime/config.go:98]: It is more correct for checking omitted state with nil pointer in Go, please leave as is. Doing all at once will not be reviewed and merged, so best to do over time.

File: pkg/scheduler/server/api.go
Comment (JoshVanL) [inline pkg/scheduler/server/api.go]: Why do we not return an error here? We need at least the name of the job in the log line.

Comment (JoshVanL) [reply pkg/scheduler/server/api.go]: This seems to happen when the metadata does not have all the correct fields. Please move this higher in a if block which checks if any of the fields are empty.

File: pkg/scheduler/server/api.go
Comment (JoshVanL) [inline pkg/scheduler/server/api.go:116]:
```suggestion
```

File: pkg/scheduler/server/api.go
Comment (JoshVanL) [inline pkg/scheduler/server/api.go:120]: Why are we no longer returning error if the job did not exist? An error code on deleting an object which did not exist is highly useful and is what would be expected by the user.

File: pkg/scheduler/server/api.go
Comment (JoshVanL) [inline pkg/scheduler/server/api.go:166]: Why are we truncating the time?

File: pkg/scheduler/server/api.go
Comment (JoshVanL) [inline pkg/scheduler/server/api.go:164]:
```suggestion
		return nil, err
```

File: pkg/scheduler/server/api.go
Comment (JoshVanL) [inline pkg/scheduler/server/api.go:172]:
```suggestion
		return -1, nil
```

File: pkg/scheduler/server/api.go
Comment (JoshVanL) [inline pkg/scheduler/server/api.go:177]:
```suggestion
		return -1, fmt.Errorf("parse error: %w", err)
```

File: pkg/scheduler/server/api.go
Comment (JoshVanL) [inline pkg/scheduler/server/api.go:175]: Please can we make `timeutils.ParseDuration(ttl)` return a `struct`. Returning a list of similar typed variables is a source of bugs.

File: pkg/scheduler/server/api.go
Comment (JoshVanL) [inline pkg/scheduler/server/api.go:107]: Where is the authorization that a client can perform this request done? i.e. the client resides in the same namespace.

File: pkg/scheduler/server/server.go
Comment (JoshVanL) [reply pkg/scheduler/server/server.go:156]: If `runEtcdCron` encapsulates running both etcd and cron then this name is more correct, though I'd ask why these two long running processes are not distinct with `runETCD` and `runCron`.

File: pkg/scheduler/server/server.go
Comment (JoshVanL) [inline pkg/scheduler/server/server.go:226]: Question: why are we using a builder style initialization rather than the more consistent `Options`?

File: pkg/scheduler/server/server.go
Comment (JoshVanL) [inline pkg/scheduler/server/server.go:241]: We should also be waiting when etcd returns an error.

File: tests/integration/framework/util/parallel.go
Comment (JoshVanL) [reply tests/integration/framework/util/parallel.go]: If errors are correctly being reported when a parallel test fails then it can be safely removed without replacement. Please delete the code instead of commenting it out.

File: tests/integration/suite/scheduler/api/http/errors.go
Comment (JoshVanL) [reply tests/integration/suite/scheduler/api/http/errors.go:259]: Why are we making this design decision?

File: tests/integration/suite/scheduler/api/jobs.go
Comment (JoshVanL) [inline tests/integration/suite/scheduler/api/jobs.go]:
```suggestion
```

File: tests/integration/suite/scheduler/api/jobs.go
Comment (JoshVanL) [inline tests/integration/suite/scheduler/api/jobs.go]:
```suggestion
func getEtcdKeys(t *testing.T, ctx context.Context, port int) []*mvccpb.KeyValue {
```

**Review by JoshVanL — CHANGES_REQUESTED** [review]

File: pkg/actors/actors.go
Comment (artursouza) [reply pkg/actors/actors.go:1217]: actors are not bound to an appId - the same actor type can be served by multiple appIds. The reminder is bound to a given ActorType+ActorID.

File: dapr/proto/runtime/v1/dapr.proto
Comment (artursouza) [reply dapr/proto/runtime/v1/dapr.proto:199]: Dapr does not offer this scenario. Also, the Get job is not as simple in a distributed scheduler since a job is not being executed on every instance. Reading from etcd does not mean the job is being executed either. We don't offer GetJob in reminders, for example.

File: pkg/actors/actors.go
Comment (artursouza) [reply pkg/actors/actors.go:1217]: See above how the jobName is being composed.

File: dapr/proto/runtime/v1/dapr.proto
Comment (artursouza) [reply dapr/proto/runtime/v1/dapr.proto:199]: I will add Get but List needs a scenario to understand what we are solving for. I know we did not solve for observability yet.

File: pkg/scheduler/server/api.go
Comment (artursouza) [reply pkg/scheduler/server/api.go]: There will be more conditions here, so combining multiple ones early can be bug-prone and duplicate the logic on future "if" conditions.

File: pkg/scheduler/server/api.go
Comment (artursouza) [inline pkg/scheduler/server/api.go]: Handle namespace inside jobName.

File: pkg/scheduler/server/api.go
Comment (artursouza) [reply pkg/scheduler/server/api.go:107]: OK. Will see where it is for scheduler.

File: pkg/scheduler/server/api.go
Comment (artursouza) [reply pkg/scheduler/server/api.go:116]: Why not log error?

File: pkg/scheduler/server/api.go
Comment (artursouza) [reply pkg/scheduler/server/api.go:120]: Because we are not getting the job first, we call delete directly. Calling Get first is not really useful because it is not an atomic operation. Second, we removed Get API at this iteration. I am adding Get back but it should not be used here IMO.

File: pkg/scheduler/server/api.go
Comment (artursouza) [reply pkg/scheduler/server/api.go:164]: nil is not a valid return for time.Time.

File: pkg/scheduler/server/api.go
Comment (artursouza) [reply pkg/scheduler/server/api.go:166]: Because it is the finest granularity of the jobs.

File: pkg/scheduler/server/api.go
Comment (artursouza) [reply pkg/scheduler/server/api.go:175]: Yes, we will. I am using the Dapr code as-is for now.

File: pkg/scheduler/server/api.go
Comment (artursouza) [reply pkg/scheduler/server/api.go:172]: We treat 0 the same way to negative in the library. It ignores TTL in that case.

File: pkg/scheduler/server/api.go
Comment (artursouza) [reply pkg/scheduler/server/api.go:177]: Same. 0 means not-set.

File: pkg/scheduler/server/server.go
Comment (artursouza) [reply pkg/scheduler/server/server.go:226]: This is the library's original code. We can refactor that at some point.

File: pkg/scheduler/server/server.go
Comment (artursouza) [reply pkg/scheduler/server/server.go:241]: Let me see if I can add that. I am just changing the cron lib use to match the latest version.

File: tests/integration/framework/util/parallel.go
Comment (artursouza) [reply tests/integration/framework/util/parallel.go]: I don't know if they are because there is nothing connecting t with collections anymore.

Comment (artursouza) [reply tests/integration/framework/util/parallel.go]: This is unrelated to my change I just saw this warning.

File: tests/integration/suite/scheduler/api/http/errors.go
Comment (artursouza) [reply tests/integration/suite/scheduler/api/http/errors.go:259]: Same as before: get and delete are not transaction, so there is no guarantee by doing the get upfront. Also, etcd accepts the delete as successful even if key does not exist.

File: pkg/actors/actors.go
Comment (artursouza) [reply pkg/actors/actors.go:1217]: Updated to new jobName composition.

File: dapr/proto/runtime/v1/dapr.proto
Comment (artursouza) [reply dapr/proto/runtime/v1/dapr.proto:199]: Added Get back.

File: pkg/scheduler/server/api.go
Comment (artursouza) [reply pkg/scheduler/server/api.go]: Added metadata to the output since callback does not have jobName (yet).

File: pkg/scheduler/server/api.go
Comment (artursouza) [reply pkg/scheduler/server/api.go:107]: This is not implemented yet. Adding a TODO.

File: pkg/scheduler/server/api.go
Comment (artursouza) [reply pkg/scheduler/server/api.go]: Moved namespace to metadata.

Comment (artursouza) [reply pkg/scheduler/server/api.go]: Also, concatenation is done 100% in Dapr.

File: pkg/scheduler/server/server.go
Comment (artursouza) [reply pkg/scheduler/server/server.go:241]: There is a defer on etcd close() already.

**Comment by artursouza** [comment]
This passes now: `go test -v -race -tags integration ./tests/integration -focus scheduler` 

```
    --- PASS: Test_Integration/scheduler/api/grpc/basic (3.45s)
        --- PASS: Test_Integration/scheduler/api/grpc/basic/run (3.35s)
            --- PASS: Test_Integration/scheduler/api/grpc/basic/run/bad_request (0.00s)
            --- PASS: Test_Integration/scheduler/api/grpc/basic/run/good_request (0.09s)
    --- PASS: Test_Integration/scheduler/api/http/basic (3.53s)
        --- PASS: Test_Integration/scheduler/api/http/basic/run (3.40s)
            --- PASS: Test_Integration/scheduler/api/http/basic/run/bad_json (0.01s)
            --- PASS: Test_Integration/scheduler/api/http/basic/run/good_json (0.13s)
    --- PASS: Test_Integration/scheduler/api/jobs (3.79s)
        --- PASS: Test_Integration/scheduler/api/jobs/run (3.70s)
            --- PASS: Test_Integration/scheduler/api/jobs/run/CRUD_10_jobs (0.45s)
    --- PASS: Test_Integration/scheduler/quorum/notls (6.72s)
        --- PASS: Test_Integration/scheduler/quorum/notls/run (5.53s)
PASS
ok      github.com/dapr/dapr/tests/integration  18.643s
```

**Comment by artursouza** [comment]
I am merging this to unblock @cicoyle - it is far from being the last PR. I will pause sending PRs for now until I am done with the perf tests.

**Pull request merged**
